### PR TITLE
fix: constructor validation on build

### PR DESCRIPTION
**Pull request by osalkanovic**
#224

I have commented unordered map constructor to simulate issue and I expect to CI fail.
When someone review PR I will return previous state of unordered map :)

### CODE REVIEW

**Comment by osalkanovic** [comment]
Output from CI: image

**Comment by volovyks** [comment]
@osalkanovic you can add tests that can expect something to fail or not to fail, check examples here: https://github.com/near/near-sdk-js/blob/develop/tests/__tests__/decorators/payable.ava.js

The fact that the constructor exists does not mean that all the variables were initialized in it. So users can get errors anyway.

**Comment by osalkanovic** [comment]
@ailisp @volovyks its ready for review. 
Now we have covered all cases.
Idk why I have linter issue 🥲

**Comment by osalkanovic** [comment]
And ya,
ts-morph is amazing tool we can cover other cases if necessary

File: tests/src/constructor-validation/version-4.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { NearBindgen, LookupMap } from "near-sdk-js";
```
Comment (volovyks) [inline tests/src/constructor-validation/version-4.ts:1]: nit: let's name these files according to what they are. Like: `NoConstructorContract.ts`

File: cli/contract_validation.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      const hasConstructor = constructors.length > 0;
+      const propertiesToBeInited = properties.filter((p) => !p.initializer);
+      // reson: examples/clean-state.js
+      if (!hasConstructor && propertiesToBeInited.length === 0) {
+        return true;
+      }
+      if (!hasConstructor && propertiesToBeInited.length > 0) {
+        console.log(
+          chalk.redBright(
+            `Ops, constructor isnt initialized, after initialization include ${propertiesToBeInited
```
Comment (volovyks) [inline cli/contract_validation.js:28]: We are not initializing the constructor, we are initializing parameters inside the constructor.
I think something like "All parameters must be initialized in the constructor" is closer to what we want to say here.

**Review by volovyks — COMMENTED** [review]
@osalkanovic wow, it's a great approach! I think this solution is great. Let's address a few comments and merge it.

P.S. Just run `yarn lint` and `yarn format` before the commit. It will fix most of the linter errors.

File: tests/__tests__/constructor_validation.ava.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      "near-sdk-js build src/constructor-validation/version-3.ts build/constructor-validation/version-3.wasm"
+    ).toString();
+  } catch (e) {
+    result = e;
+  }
+
+  t.is(result.status, 2);
+});
+
+test("should throw error, construcor isnt declared", async (t) => {
```
[the remark this answers is not in the document]
Comment (ailisp) [reply tests/__tests__/constructor_validation.ava.js:42]: I think result.stdout / result.stderr records the error text

File: cli/cli.js [hunk 4]
```diff
@@ -86,6 +86,7 @@ async function checkTsBuildWithTsc(sourceFileWithPath) {
 
 // Common build function
 async function createJsFileWithRullup(sourceFileWithPath, rollupTarget) {
+  await validateContract(sourceFileWithPath);
```
Comment (ailisp) [inline cli/cli.js:89]: Nice step! All existing contracts can pass the validation

**Comment by osalkanovic** [comment]
> Great work! Besides @volovyks 's comprehensive comments, it looks good to me.

@volovyks @ailisp I will update PR today

**Comment by volovyks** [comment]
Hey @osalkanovic ! This PR is getting outdated, it will be hard to continue your work on it if we postpone it. Do you have the capacity to continue doing this?

File: cli/contract_validation.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  for (const contractClass of contractClasses) {
+    const classStructure = contractClass.getStructure();
+    const { decorators, properties } = classStructure;
+    const hasBindgen = decorators.find(
+      (decorator) => decorator.name === "NearBindgen"
+    );
+    if (hasBindgen) {
+      const constructors = contractClass.getConstructors();
+      const hasConstructor = constructors.length > 0;
+      const propertiesToBeInited = properties.filter((p) => !p.initializer);
```
[the remark this answers is not in the document]
Comment (volovyks) [reply cli/contract_validation.js]: Do not work for JS contracts due to the absence of parameter declaration.

File: tests/__tests__/constructor_validation.ava.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (volovyks) [reply tests/__tests__/constructor_validation.ava.js]: The validation error is overridden with another top-level error. Added const for now.

File: tests/__tests__/constructor_validation.ava.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import test from "ava";
+import { execSync } from "child_process";
+
+test("should build, constructor is correctly initialized", async (t) => {
+  let result = null;
+  try {
+    result = execSync(
```
[the remark this answers is not in the document]
Comment (volovyks) [reply tests/__tests__/constructor_validation.ava.js]: done

File: tests/__tests__/constructor_validation.ava.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  } catch (e) {
+    result = e;
+  }
+
+  t.is(result.status, 2);
+});
+
+test("should throw error, construcor is empty", async (t) => {
+  let result = null;
+  try {
```
[the remark this answers is not in the document]
Comment (volovyks) [reply tests/__tests__/constructor_validation.ava.js]: done

File: tests/__tests__/constructor_validation.ava.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+test("should build, constructor is correctly initialized", async (t) => {
+  let result = null;
+  try {
+    result = execSync(
+      "near-sdk-js build src/constructor-validation/version-1.ts build/constructor-validation/version-1.wasm"
+    ).toString();
+  } catch (e) {
+    result = e;
+  }
+  t.not(result.status, 2);
```
[the remark this answers is not in the document]
Comment (volovyks) [reply tests/__tests__/constructor_validation.ava.js]: const created

**Review by volovyks — APPROVED** [review]
@ailisp I have refactored the code. Rewrote everything in TS, moved contract validation to the top level, and rewrote tests. Let's merge it if you are ok with my changes.

**Comment by ailisp** [comment]
@volovyks Looks great :+1: The original functionality by @osalkanovic is good, and your refactors, rewrote tests and in TS makes the code quality better than I expected. Merge now

**Comment by volovyks** [comment]
@osalkanovic thanks for the help!

**Pull request merged**